File: estimation_tool.py
import hashlib
import os
import csv
import fnmatch

def calculate_sha3_256_hash(file_path):
    hasher = hashlib.sha3_256()
    with open(file_path, 'rb') as file:
        buf = file.read(65536)
        while len(buf) > 0:
            hasher.update(buf)
            buf = file.read(65536)
    return hasher.hexdigest()

def count_code_lines(file_path):
    code_lines = 0
    in_block_comment = False
    try:
        with open(file_path, 'r', encoding='utf-8', errors='ignore') as file:
            for line in file:
                stripped_line = line.strip()
                if stripped_line.startswith("/*"):
                    in_block_comment = True
                if in_block_comment and "*/" in stripped_line:
                    in_block_comment = False
                    stripped_line = stripped_line.split("*/")[-1]
                if not in_block_comment and not stripped_line.startswith("//") and stripped_line:
                    code_lines += 1
    except Exception as e:
        print(f"Error processing file {file_path}: {e}")
    return code_lines

def make_relative_path(path, start):
    rel_path = os.path.relpath(path, start=start)
    if not rel_path.startswith('.'):
        rel_path = f"./{rel_path}"
    return rel_path

def is_valid_file(file_name):
    valid_extensions = ['.sol', '.vy', '.rs']
    excluded_patterns = ['*.t.sol', '*.s.sol']

    for pattern in excluded_patterns:
        if fnmatch.fnmatch(file_name, pattern):
            return False

    return any(file_name.endswith(ext) for ext in valid_extensions)

def main(target_folder):
    base_folder = os.path.abspath(os.path.join(target_folder, os.pardir))
    output_filename = os.path.basename(os.path.normpath(target_folder)) + '_hash.csv'
    output_file = os.path.join(base_folder, output_filename)

    with open(output_file, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(['File Path & SHA-3 Hash', 'Code Lines'])

        for root, dirs, files in os.walk(target_folder):
            for file in files:
                if not is_valid_file(file):
                    continue
                file_path = os.path.join(root, file)
                relative_file_path = make_relative_path(file_path, base_folder)
                hash_value = calculate_sha3_256_hash(file_path)
                code_lines = count_code_lines(file_path)
                combined_info = f"Path: {relative_file_path}\nSHA3: {hash_value}"
                writer.writerow([combined_info, code_lines])
                print(f"Processed file: {relative_file_path}")

if __name__ == "__main__":
    target_folder = input("Enter the path of the folder to count: ")
    main(target_folder)

File: estimation_tool.ts
import fs from 'fs-extra';
import path from 'path';
import crypto from 'crypto';
import * as glob from 'glob';

const validExtensions = ['.sol', '.vy', '.rs', '.move'];
const excludedPatterns = ['.*\\.t\\.sol$', '.*\\.s\\.sol$'];

function calculateSHA3_256Hash(fileContent: Buffer): string {
    return crypto.createHash('sha3-256').update(fileContent).digest('hex');
}

function countCodeLines(fileContent: string): number {
    const lines = fileContent.split('\n');
    let codeLines = 0;
    let inBlockComment = false;

    lines.forEach((line) => {
        const trimmedLine = line.trim();

        if (trimmedLine.startsWith('/*')) {
            inBlockComment = true;
        }
        if (inBlockComment && trimmedLine.endsWith('*/')) {
            inBlockComment = false;
            return;
        }
        if (!inBlockComment && !trimmedLine.startsWith('//') && trimmedLine !== '') {
            codeLines++;
        }
    });

    return codeLines;
}

function isExcluded(file: string): boolean {
    return excludedPatterns.some((pattern) => new RegExp(pattern).test(file));
}

function formatCsvRow(filePath: string, hash: string, codeLines: number): string[] {
    const firstColumn = `Path: ${filePath}\nSHA3: ${hash}`;
    return [`"${firstColumn}"`, codeLines.toString()];
}

function processFiles(targetFolder: string, outputFile: string): void {
    const allFiles = glob.sync('**/*', { cwd: targetFolder, nodir: true });

    const data: string[] = ['File Path & SHA-3 Hash, Code Lines'];

    allFiles.forEach((file) => {
        const fullPath = path.join(targetFolder, file);

        if (isExcluded(file) || !validExtensions.includes(path.extname(file))) {
            return;
        }

        try {
            const relativePath = `./${file}`;

            const fileContent = fs.readFileSync(fullPath);
            const hash = calculateSHA3_256Hash(fileContent);
            const codeLines = countCodeLines(fileContent.toString());

            const csvRow = formatCsvRow(relativePath, hash, codeLines);

            data.push(csvRow.join(', '));
            console.log(`Processed file: ${file}`);
        } catch (error) {
            console.error(`Error processing file ${file}: ${error}`);
        }
    });

    fs.writeFileSync(outputFile, data.join('\n') + '\n');
}

const targetFolder = process.argv[2];
const baseFolder = path.dirname(targetFolder);
const outputFilename = `${path.basename(targetFolder)}_hash.csv`;
const outputFile = path.join(baseFolder, outputFilename);

processFiles(targetFolder, outputFile);
